### PR TITLE
Upgraded TSM Loops function to use MG

### PR DESCRIPTION
**Pull request by cpviolator**
This commit upgrades the Loops TSM routine to enable the Multigrid solver for stochastic sources. This branch removes the stochastic Even/Odd deflation options and simply solves the stochastic source with a direct, preconditioned MG solver. The full spinor is then reconstructed, and the exact part is projected out from the solution, leaving the inexact part.

### CODE REVIEW

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: The comment in lines 8716-8718 is not fully correct. Truncated Solver Method (TSM) is not directly related to projecting out the exact contribution from the solution of the stochastic inversion. Rather, TSM can just be employed in the stochastic part as an additional measure of reducing computer time. In other words, projecting out the exact contribution from the solution is _always_ required by the exact-eigenvalue-deflation method itself, regardless of the TSM being employed or not. I suggest that the last sentence (lines 8716-8718) is removed.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: maybe we should set "out" to zero here, to make sure that initial guess is zero for the inversion, i.e.:
zeroCuda(*out);

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: Maybe we should set "out" and "out_LP" to zero, before the "dirac.prepare(....)" to make sure that initial guess is zero for the inversion, i.e.:
zeroCuda(*out);
zeroCuda(*out_LP);

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: I believe that the smethod option is redundant in the way that the code is constructed right now. I suggest that we just remove it from everywhere.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: The if(smethod==0) statement should be removed. There's no other case than smethod=0.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: The following is a misleading comment. It's not any exact part that is projected out of the solution. The solution itself (the spinor "x") gets projected with (1-P), as the original comment stated.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: line 9131: It should be: (*solve_LP)(*out_LP,*in);
I.e. replace "out" with "out_LP"

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: Again, the following is a misleading comment. It's not any exact part that is projected out of the solution. The solution itself (the spinor "x") gets projected with (1-P), as the original comment stated.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: Again, the following is a misleading comment. It's not any exact part that is projected out of the solution. The solution itself (the spinor "x_LP") gets projected with (1-P), as the original comment stated.

File: lib/interface_quda.cpp
Comment (ckallidonis) [inline lib/interface_quda.cpp]: The if(smethod==0) statement is redundant and should be removed.

File: qkxtm/CalcMG_Loops_w_oneD_TSM_wExact.cpp
Comment (ckallidonis) [inline qkxtm/CalcMG_Loops_w_oneD_TSM_wExact.cpp]: We should remove the "smethod" option completely.

File: qkxtm/QKXTM_util.cpp
Comment (ckallidonis) [inline qkxtm/QKXTM_util.cpp]: We should remove smethod option completely.

**Review by ckallidonis — COMMENTED** [review]
The comment on line-9131 of lib/interface_quda.cpp should be addressed. Other than that I believe that the code should work fine. But I suggest that my other comments are also addressed, in order to remove unnecessary pieces, and to avoid confusions in the future.

File: lib/interface_quda.cpp
Comment (cpviolator) [reply lib/interface_quda.cpp]: That is an incredible catch!! Thank you.

**Review by ckallidonis — DISMISSED** [review]

**Comment by ckallidonis** [comment]
The above commits were required for a smooth compilation of the code.

**Comment by cpviolator** [comment]
Hi @ckallidonis,

The `common.h` and `cblas.h` includes are necessary for compilation with magma-1.7.0 built against OpenBlas. Please leave them in. If one prefers to build with magma built against MKL, then the `mkl.h` header should be included. This is an issue that needs to be fixed, and is related to Intel compilers being used to compile QUDA mainline: https://github.com/lattice/quda/issues/566. We should try to fix the issue given that we have working Intel builds.

Of course, one can build magma-1.7.0 against MKL with GNU threads, but MKL comes with a host of library dependencies `-lmkl_gf_lp64 -lmkl_gnu_thread -lmkl_core -lpthread -lstdc++ -lm -lgfortran` whereas an OpenBlas build just has one `-lopenblas`. This issue is very fixable, but would require some editing of the CMakeLists.txt file so that the user can specify which kind of magma build one has, and then open up the options for MKL or OpenBlas. Then, preprocessor directives can be defined and used:

```c++
#ifdefine HAVE_MKL
#include mkl.h
#endif
#ifdefine HAVE_OPENBLAS
#include common.h
#include cblas.h
#endif
```
Passing explicit paths to dependences without using CMake is confusing for new users, and should be avoided.

**Review by cpviolator — COMMENTED** [review]
Hi @ckallidonis,

The `common.h` and `cblas.h` includes are necessary for compilation with magma-1.7.0 built against OpenBlas. Please leave them in. If one prefers to build with magma built against MKL, then the `mkl.h` header should be included. This is an issue that needs to be fixed, and is related to Intel compilers being used to compile QUDA mainline: https://github.com/lattice/quda/issues/566. We should try to fix the issue given that we have working Intel builds.

Of course, one can build magma-1.7.0 against MKL with GNU threads, but MKL comes with a host of library dependencies `-lmkl_gf_lp64 -lmkl_gnu_thread -lmkl_core -lpthread -lstdc++ -lm -lgfortran` whereas an OpenBlas build just has one `-lopenblas`. This issue is very fixable, but would require some editing of the CMakeLists.txt file so that the user can specify which kind of magma build one has, and then open up the options for MKL or OpenBlas. Then, preprocessor directives can be defined and used:

```c++
#ifdefine HAVE_MKL
#include mkl.h
#endif
#ifdefine HAVE_OPENBLAS
#include common.h
#include cblas.h
#endif
```
Passing explicit paths to dependences without using CMake is confusing for new users, and should be avoided.

**Comment by ckallidonis** [comment]
Is it easy for you to do this, since you are more experienced with CMake, etc?

**Comment by ckallidonis** [comment]
By the way, the routine which performs the loops gives a seg-fault, that's why I haven't merged yet. I'm trying to solve it and then report back.

**Comment by cpviolator** [comment]
Sure, I could change the CMake, but at the moment this repo is configured to use magma compiled against OpenBlas. That's why the OpenBlas headers are needed. An Intel build requires the user to manually change the `CUDA_HOST_COMPILER` flag to `g++`, and a magma dependency built against MKL requires adding the extra libraries to the path via CMake. For the moment, please revert to the OpenBlas headers. Once I've fixed QUDA mainline to automatically detect an Intel build so CMake can change `CUDA_HOST_COMPILER`, I'll add options for MKL and Intel in this repo with the above preprocessor flags.

**Comment by ckallidonis** [comment]
But that could very inconvenient, because if the headers are reverted, every time that the affected files are included in a commit this issue will appear, because otherwise the code cannot be compiled with Intel, i.e. one has to change these includes every time the files change, before pushing the commit on Github. Is there a more convenient solution for now?

**Comment by cpviolator** [comment]
Sure, compile with GNU and OpenBlas. I can show you how to edit CMake to open up options for MKL and Intel. The Intel build you have is not documented and difficult to reproduce without prior knowledge of the `CUDA_HOST_COMPILER` issue, so until its fixed this repo should remain stable by using OpenBlas and GNU. Feel free to create a new branch that has MKL and Intel dependency structure. Once I've shown you how to edit CMake to open up the options, everything can be merged after review.

Until then, please revert to the OpenBlas headers.

**Comment by ckallidonis** [comment]
All right. But first, let me make the code work correctly.

**Comment by ckallidonis** [comment]
If it's possible, please do not make any mods to the code until I push the corrected Loops-part on Github. As soon as this is on Github, I will switch the header files back to OpenBlas and commit-push again. Then we can discuss how to make the options for MKL-Intel and GNU-OpenBlas build in another branch. In the meantime, I will develop the HighMomForm in yet another branch, but this should be independent of the options, and we should expect no conflicts.

**Comment by cpviolator** [comment]
I dismissed your approval so no merging can take place until the seg-fault
issue is fixed (you can also reverse an approval if you find faults.) Can
you provide some information about this so we can help with debugging? It's
not a good practice to push commits to a pre approved pull request as
unreviewed changes can sneak in through the back door. Once a pull request
is approved, it should be merged and the branch deleted.

HighMomForm can be added at your leisure via a separate branch, then merged
wth master. None of the issues in this branch will be affected.

As for as the MKL/OpenBlas issue, it would be good to coordinate with QUDA
on this as working Intel builds are clearly a problem. If we can find a way
to use the same repo to build with GCC or Intel via CMake, they would love
to see it!

**Comment by ckallidonis** [comment]
Well, I thought that before you created the Pull Request you had already at least compiled the code, if not run it to see that it runs and it also produces the right results. What I have in mind is that any version of the code going to be merged with the master branch should be production-ready. In any case, in my last commit I have fixed the seg-fault that existed from the [first](https://github.com/ETMC-QUDA/quda-QKXTM-Multigrid/pull/4/commits/cf787697e526f7c863dea3a46cf5a9d84f3d2fbc) commit. From first sight it seems that I get the correct results both for correlation functions and the loops. But, before I commit and push, please let me perform a thorough check and see some more test cases.

**Comment by cpviolator** [comment]
This is a function that you asked to me to rewrite to be MG enabled. It has
been changed significantly. It is impossible for me to ascertain what
'right results' are, given that MG has very different tolerance stopping
conditions to the previous version. If you want me to run functions and
cross check data before committing, please specify agreed upon run
parameters and data and specifically ask for a cross check. We have a
review process, I followed your comments, the code compiled (with GNU and
OpenBlas.) If you find a bug, please share the information as quickly as
possible so it can be fixed more speedily. Understanding bugs from
corrected code is much harder than just sharing the information first hand.
If we cannot learn from our mistakes, we will most likely repeat them.

I'm glad you're on board with version control. The whole point of a master
branch, and pushing changes to auxiliary branches for code review and
testing, ensures that the master branch is trusted. By pushing to an
auxiliary branch, requesting a code review, and having multiple
contributors test, allows us to raise concerns. Mistakes, bugs, and
conflicts, are inevitable. Code review clears them up quickly and
collaboratively. In future, if you have issues with a pull request, do not
approve it then make your own additions. Comment on what needs to be
adjusted to the branch, and create your own branch to push your own commits
so other people can review them. If code commits were meant to be perfect
first time, we wouldn't ned a review process, so take some time to
throughly test and review the code. And again, please do not piggy back
your own commits onto another person's approved pull. This is an
administrative privilege, and should be used very sparingly.

**Comment by ckallidonis** [comment]
The above commit is a working implementation of the code, supporting two- and three-point functions and the quark loops. NOTE: For now, the Loops calculation **supports only xgrid=1 and ygrid=1**. This is due to the function which performs the Fourier transform as well as the covariant derivative for the contractions. I am working on extending this for any xgrid and ygrid. I will push another commit once it's updated and crosschecked. Then we can merge and deal with other things on separate branches. Or, we can merge as it is and make the changes in a separate branch. If you prefer the second option, you can review the changes and complete the merge.

**Comment by ckallidonis** [comment]
In the meantime, there seem to be six pointers which are not freed, as there is a Warning at the end of the execution when running for the Loops. This could be a leak and cause an out-of-memory issue, especially for large lattices or large numbers of stochastic vectors. It would be great if you could help spot this leak, as my attempts were not successful up to now.

**Review by ckallidonis — APPROVED** [review]
This code produces correct two- and three-point functions. It also runs for the disconnected loops, as long as xgrid=1 and ygrid=1.

**Comment by ckallidonis** [comment]
I will merge the two branches so as to have a working master branch and to proceed with other tasks. Intel support will be included in a separate branch, that hopefully will be parallel to the master branch with the only difference being in the affected header files.

**Comment by cpviolator** [comment]
That's great work. However, please do not merge to master before code review!! You'll notice that each time I make a change to the code, I request that it be reviewed before merging. This is an important step in team coding. Changes I make should be approved by somebody else. Changes you make should be approved by somebody else. Some 4000 lines of code were changed in commit 9cce967, then pushed directly to master with no review process. Without these important steps, we are failing to work as a team.

The Intel changes will be minor. Some CMake additions and a few preprocessor directives. There's no need for a parallel branch, just a hotfix branch to make the changes, a review process, and then merging to master after testing.

**Comment by ckallidonis** [comment]
Yes, I know that changes should be reviewed before merging from at least one team member. I just took the initiative in this case because I am sure that with the latest commit the code is fully working. Of course you can always provide your own suggestions/comments!

Regarding the Intel branch, since I'm running on Piz Daint with an Intel build, I would prefer to have a separate branch until we settle this down, such that when commits take place, I will not carry the affected header files as well into the master branch. Otherwise I will have to revert back and forth all the time. The drawback from this is that two Pull Requests will need to take place every time (one for master and one for the Intel-build), but hopefully this will not last for long.

**Comment by cpviolator** [comment]
I actually took a look at your CMake file, and there is a very simple fix.

You pass all your headers and libraries 'manually' because you keep them
all in the same place. All we need do in that case is have a simple switch
in CMake. One for MKL, one for OpenBlas. Then a preprocessor flag can be
placed around the headers, and we can stay on a single master branch. What
do you think?

BTW, I'm using the plug-in to compile and run Simone's recent work on
\delta_mu at hackathon. I'll share results once I get them, and we can talk
about implementing his code if it's faster
https://github.com/lattice/quda/tree/hotfix/mu_factor

**Comment by ckallidonis** [comment]
Yes, this sounds like a plan. Feel free to do the switch in the way you believe it's best.

Regarding Simone's code, you mean we see what parts are affected and then we manually implement the additional code in this repository instead of merging, etc? If yes, this could be an option too.

**Comment by cpviolator** [comment]
That's the benefit of having a plug-in. We can just point our code to any
QUDA source code, and compile against it. We can use the plug-in to
experiment with new features without having to do the annoying work of
copying and editing files. The plug-in has a CMake variable ${QUDA_HOME}
that you use to point to new versions, then compile as normal. It's as
simple as that.

**Pull request merged**
